File: src/parser/Token.cpp
// This source file is part of the polarphp.org open source project
//
// Copyright (c) 2017 - 2019 polarphp software foundation
// Copyright (c) 2017 - 2019 zzu_softboy <zzu_softboy@163.com>
// Licensed under Apache License v2.0 with Runtime Library Exception
//
// See https://polarphp.org/LICENSE.txt for license information
// See https://polarphp.org/CONTRIBUTORS.txt for the list of polarphp project authors
//
// Created by polarboy on 2019/07/09.

#include "polarphp/parser/Token.h"
#include "llvm/Support/raw_ostream.h"
#include "polarphp/syntax/TokenKinds.h"

namespace polar::parser {

using namespace polar::syntax;

void Token::dump() const
{
   dump(llvm::errs());
   llvm::errs() << "\n";
}

/// Dump this piece of syntax recursively.
void Token::dump(raw_ostream &outStream) const
{
   outStream << "============================" << "\n";
   outStream << "name: ";
   dump_token_kind(outStream, m_kind);
   outStream << "\n";
   if (is_keyword_token(m_kind) || is_punctuator_token(m_kind)) {
      outStream << "text: " << getLexicalText() << "\n";
   }
   if (!isInvalidLexValue()) {
      if (m_kind == TokenKindType::T_VARIABLE ||
          m_kind == TokenKindType::T_IDENTIFIER_STRING ||
          m_kind == TokenKindType::T_STRING_VARNAME) {
         outStream << "value: ";
         if (m_kind == TokenKindType::T_VARIABLE){
            outStream << '$';
         }
         outStream << getValue<std::string>() << "\n";
      } else if (m_kind == TokenKindType::T_LNUMBER) {
         outStream << "value: ";
         outStream << getValue<std::int64_t>() << "\n";
      } else if (m_kind == TokenKindType::T_DNUMBER) {
         outStream << "value: ";
         outStream << getValue<double>() << "\n";
      } else if (m_kind == TokenKindType::T_CONSTANT_ENCAPSED_STRING ||
                 m_kind == TokenKindType::T_ENCAPSED_AND_WHITESPACE) {
         const std::string &text = getValue<std::string>();
         outStream << "length: " << text.size() << "\n";
         outStream << "value: " << text << "\n";
      } else if (m_kind == TokenKindType::T_ERROR && hasValue()) {
         outStream << "error: ";
         outStream << getValue<std::string>() << "\n";
      }
   } else {
      outStream << "value: invalid lex value" << "\n";
   }
}

} // polar::parser

File: devtools/gyb/src/Syntax/Token.php
<?php
// This source file is part of the polarphp.org open source project
//
// Copyright (c) 2017 - 2019 polarphp software foundation
// Copyright (c) 2017 - 2019 zzu_softboy <zzu_softboy@163.com>
// Licensed under Apache License v2.0 with Runtime Library Exception
//
// See https://polarphp.org/LICENSE.txt for license information
// See https://polarphp.org/CONTRIBUTORS.txt for the list of polarphp project authors
//
// Created by polarboy on 2019/10/28.

namespace Gyb\Syntax;

/**
 * Represents the specification for a Token in the TokenSyntax file.
 *
 * Class Token
 * @package Gyb\Syntax
 */
class Token
{
   /**
    * @var string $name
    */
   private $name;
   /**
    * @var string $kind
    */
   private $kind;
   /**
    * @var string $serializationCode
    */
   private $serializationCode;
   /**
    * @var string $text
    */
   private $text;
   /**
    * @var string $classification
    */
   private $classification;
   /**
    * @var bool $keyword
    */
   private $keyword;

   /**
    * @var string $valueType
    */
   private $valueType = '';

   public function __construct(string $name, string $kind, int $serializationCode,
                               string $text = '', string $valueType = '', string $classification = 'None',
                               bool $keyword = false)
   {
      $this->name = trim($name);
      $this->kind = trim($kind);
      $this->serializationCode = $serializationCode;
      $this->text = trim($text);
      $this->classification = SyntaxClassification::getByName(trim($classification));
      $this->keyword = $keyword;
      $this->valueType = $valueType;
   }

   public function getPolarKind(): string
   {
      $name = lcfirst($this->name);
      if ($this->keyword) {
         return $name . 'Keyword';
      }
      return $name;
   }

   /**
    * @return mixed
    */
   public function getRawName()
   {
      return $this->name;
   }

   public function getName()
   {
      return $this->name.'Token';
   }

   /**
    * @return mixed
    */
   public function getKind()
   {
      return $this->kind;
   }

   /**
    * @return mixed
    */
   public function getSerializationCode()
   {
      return $this->serializationCode;
   }

   /**
    * @return mixed
    */
   public function getText()
   {
      return $this->text;
   }

   /**
    * @return bool
    */
   public function isTextEmpty(): bool
   {
      return strlen($this->text) == 0;
   }

   /**
    * @return mixed
    */
   public function getClassification()
   {
      return $this->classification;
   }

   /**
    * @return bool
    */
   public function isKeyword(): bool
   {
      return $this->keyword;
   }

   /**
    * @return string
    */
   public function getMacroName(): string
   {
      return 'TOKEN';
   }

   /**
    * @return string
    */
   public function getValueType(): string
   {
      return $this->valueType;
   }
}